idx/metric_definition.go

```diff
@@ -0,0 +1,552 @@
+}
+
+// TagKeyValues stores a slice of all of the Tag Key/Value pair combinations for a MetricDefinition
+type TagKeyValues struct {
+	KeyValues KeyValuesSlice
+}
+
+// Strings returns a slice containing all of the Tag Key/Value pair combinations for a MetricDefinition.
+// Each item in the slice is in the form of 'key=value'
+func (t *TagKeyValues) Strings() []string {
+	tags := make([]string, len((*t).KeyValues))
+	for i, tag := range (*t).KeyValues {
+		tags[i] = tag.String()
+	}
+	return tags
+}
+
+func (t *TagKeyValues) ExtensionType() int8 {
+	return 90
+}
+
+func (t *TagKeyValues) Len() int {
+	var total int
+	for _, kv := range t.Strings() {
+		if strings.HasPrefix(kv, "name") {
+			continue
+		}
+		total += len(kv) + 1
+	}
+	return total
+}
+
+func (t *TagKeyValues) MarshalBinaryTo(b []byte) error {
+	var total int
+	for _, kv := range t.Strings() {
+		if strings.HasPrefix(kv, "name") {
+			continue
+		}
+		if kv == "=" || kv == "" {
+			continue
+		}
+		copy(b[total:], kv)
+		total += len(kv)
+		copy(b[total:], ";")
+		total++
+	}
+	return nil
+}
+
+func (t *TagKeyValues) UnmarshalBinary(b []byte) error {
+	if len(b) < 3 {
+		return nil
+	}
+	tags := strings.Split(string(b[:len(b)-1]), ";")
+	tmp := make([]TagKeyValue, len(tags))
+	t.KeyValues = tmp
+	for i, tag := range tags {
+		if strings.HasPrefix(tag, "name") {
+			continue
+		}
+		if tag == "=" || tag == "" {
+			log.Error("idx: Empty tag, ignoring: ", tag)
+			invalidTag.Inc()
+			continue
+		}
+		if strings.Contains(tag, ";") {
+			log.Errorf("idx: Tag %q has an invalid format, ignoring", tag)
+			invalidTag.Inc()
+			continue
+		}
+		eqPos := strings.Index(tag, "=")
+		if eqPos < 0 {
+			log.Errorf("idx: Tag %q has an invalid format, ignoring", tag)
+			invalidTag.Inc()
+			continue
+		}
+		key, err := IdxIntern.AddOrGet([]byte(tag[:eqPos]), true)
+		if err != nil {
+			log.Errorf("idx: Failed to intern tag %q, %v", tag, err)
+			internError.Inc()
+			continue
+		} else {
+			(*t).KeyValues[i].Key = key
+		}
+
+		value, err := IdxIntern.AddOrGet([]byte(tag[eqPos+1:]), true)
+		if err != nil {
+			log.Errorf("idx: Failed to intern tag %q, %v", tag, err)
+			internError.Inc()
+			continue
+		} else {
+			(*t).KeyValues[i].Value = value
+		}
+	}
+	return nil
+}
+
+// MetricDefinition stores information which identifies a single metric
+type MetricDefinition struct {
+	Id    schema.MKey
+	OrgId uint32
+	// using custom marshalling for MetricName
+	// if there is another way we should explore that
+	Name       MetricName `msg:"name,extension"`
+	Interval   int
+	Unit       string
+	mtype      mType
+	Tags       TagKeyValues `msg:"tagkeyvalues,extension"`
+	LastUpdate int64
+	Partition  int32
+}
+
+// NameWithTags returns a string version of the MetricDefinition's name with
+// all of its tags in the form of 'name;key1=value1;key2=value2;key3=value3'
+func (md *MetricDefinition) NameWithTags() string {
+	var bld strings.Builder
+
+	md.Name.string(&bld)
+	sort.Sort(md.Tags.KeyValues)
```

Comment:
this sorting may lead to a lot of calls to `IdxIntern.GetStringFromPtr()` when comparing the values to each other. 2 lines down these strings get retrieved from the interning anyway. it would be more efficient to first retrieve the strings, then sort them as strings, and then loop over them.

Reply:
Sort of, but it doesn't make much of a difference and causes more allocations. Probably because I have to instead create two slices of strings in which to store the keys and values beforehand, then populate them.

Original benchmark
```
BenchmarkGetNameWithTags-12      1000000              1644 ns/op             256 B/op          5 allocs/op
```

With modified code to retrieve values and then sort
```
BenchmarkGetNameWithTags-12      1000000              1153 ns/op             416 B/op          7 allocs/op
```

For now I'm going to leave it alone, but if it becomes a hotspot later on I can revisit it.